pkg/cvo/status.go

```diff
@@ -555,7 +566,38 @@ func convertErrorToProgressing(now time.Time, statusFailure error) (reason strin
 	case payload.UpdateEffectReport:
 		return uErr.Reason, uErr.Error(), false
 	case payload.UpdateEffectNone:
-		return uErr.Reason, fmt.Sprintf("waiting on %s", uErr.Name), true
+		var exceeded []string
+		names := uErr.Names
+		if len(names) == 0 {
+			names = []string{uErr.Name}
+		}
+		var machineConfig bool
+		for _, name := range names {
+			m := 30 * time.Minute
+			// It takes longer to upgrade MCO
+			if name == "machine-config" {
+				m = 3 * m
+			}
+			t := payload.COUpdateStartTimesGet(name)
+			if (!t.IsZero()) && t.Before(now.Add(-(m))) {
+				if name == "machine-config" {
+					machineConfig = true
+				} else {
+					exceeded = append(exceeded, name)
+				}
+			}
+		}
+		// returns true in those slow cases because it is still only a suspicion
+		if len(exceeded) > 0 && !machineConfig {
+			return uErr.Reason, fmt.Sprintf("%swaiting on %s over 30 minutes which is longer than expected", slowCOUpdatePrefix, strings.Join(exceeded, ", ")), true
+		}
+		if len(exceeded) > 0 && machineConfig {
+			return uErr.Reason, fmt.Sprintf("%swaiting on %s over 30 minutes and machine-config over 90 minutes which is longer than expected", slowCOUpdatePrefix, strings.Join(exceeded, ", ")), true
+		}
+		if len(exceeded) == 0 && machineConfig {
+			return uErr.Reason, fmt.Sprintf("%swaiting on machine-config over 90 minutes which is longer than expected", slowCOUpdatePrefix), true
+		}
+		return uErr.Reason, fmt.Sprintf("waiting on %s", strings.Join(names, ", ")), true
```

Comment:
One thing I'd consider is to extract the case body into a separate method (and do the same for the `UpdateEffectFailAfterInterval` case body) so that the the high-level "what error is this" is separate from "how is this error type handled", and instead of one big method we'd have one small high-level one, and two smaller low-level ones.

Reply:
Done.